flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailoverResultPartition.java

```diff
@@ -17,41 +17,14 @@
 
 package org.apache.flink.runtime.executiongraph.failover.flip1;
 
-import org.apache.flink.runtime.io.network.partition.ResultPartitionType;
+import org.apache.flink.runtime.executiongraph.IntermediateResultPartition;
 import org.apache.flink.runtime.jobgraph.IntermediateResultPartitionID;
+import org.apache.flink.runtime.scheduler.strategy.ExecutionVertexID;
+import org.apache.flink.runtime.topology.Result;
 
 /**
- * A connection between {@link FailoverVertex FailoverVertices}.
- *
- * <p>producer -> ResultPartition -> consumer
+ * Represents a {@link IntermediateResultPartition} produced by a {@link FailoverVertex}.
  */
-public interface FailoverEdge {
-
-	/**
-	 * Returns the ID of the result partition that the source produces.
-	 *
-	 * @return ID of the result partition that the source produces
-	 */
-	IntermediateResultPartitionID getResultPartitionID();
-
-	/**
-	 * Returns the {@link ResultPartitionType} of the produced result partition.
-	 *
-	 * @return type of the produced result partition
-	 */
-	ResultPartitionType getResultPartitionType();
-
-	/**
-	 * Returns the source vertex, i.e., the producer of the result partition.
-	 *
-	 * @return source vertex
-	 */
-	FailoverVertex getSourceVertex();
-
-	/**
-	 * Returns the target vertex, i.e., the consumer of the result partition.
-	 *
-	 * @return target vertex
-	 */
-	FailoverVertex getTargetVertex();
+public interface FailoverResultPartition<V extends FailoverVertex<V, R>, R extends FailoverResultPartition<V, R>>
```

Comment:
Why does `FailoverResultPartition` need to be generic? In the further commits, it looks as if we always say `FailoverTopology<?, ?>`. Hence, we could simply say `FailoverResultPartition extends Result<ExecutionVertexID, IntermediateResultPartitionID, FailoverVertex, FailoverResultPartition>`. That way this interface does not need to be generic. The same applies to the other interfaces as well.

Reply:
Agreed that changing it like this can make it much neater.
The problem is that we cannot have one implement for both `SchedulingToplogy` and `FailoverTopology`, since type `Topology<ExecutionVertexID, IntermediateResultPartitionID, SchedulingExecutionVertex, SchedulingResultPartition>` and `Topology<ExecutionVertexID, IntermediateResultPartitionID, FailoverVertex, FailoverResultPartition>` are not compatible.
Or maybe there is a way I have not come up with yet.

An alternative is to merge `SchedulingToplogy` and `FailoverTopology` into one. The side effect is that who uses `FailoverTopology` previously would be able to see some extra information, including the vertex state, the result state and inputDependencyConstraint. It's problematic at the moment since they are all read-only interfaces.

Reply:
An example of the compatible issue with the generics simplification can be found [here](https://github.com/zhuzhurk/flink/commits/FLINK_14453_region_building_topology_test).

Reply:
@tillrohrmann should we keep the generics for FailoverTopology/SchedulingTopology? take the alternative? Or do you have any other ideas?

Reply:
I'll take a look later today.

Reply:
You are right @zhuzhurk. In order to let the `ExecutionGraphToSchedulingTopologyAdapter` implement both topologies, we need the two generic type parameters. Thanks for your explanation.